prospero-common/src/main/java/org/wildfly/prospero/galleon/ChannelMavenArtifactRepositoryManager.java

```diff
@@ -59,70 +57,83 @@ public void resolve(MavenArtifact artifact) throws MavenUniverseException {
                             artifact.getClassifier(), null);
                 } else {
                     throw new MavenUniverseException("Unable to resolve " + artifact);
                 }
             }
-            artifact.setVersion(result.getVersion());
-            artifact.setPath(result.getFile().toPath());
+
+            MavenArtifactMapper.resolve(artifact, result);
         } catch (UnresolvedMavenArtifactException e) {
             throw new MavenUniverseException(e.getLocalizedMessage(), e);
         }
     }
 
-    private DefaultArtifact streamToArtifact(Stream s) {
-        return new DefaultArtifact(s.getGroupId(), s.getArtifactId(), "jar", s.getVersion());
-    }
+    @Override
+    public void resolveAll(Collection<MavenArtifact> artifacts) throws MavenUniverseException {
+        final MavenArtifactMapper mapper = new MavenArtifactMapper(artifacts);
+        if (manifest == null) {
+            final List<org.wildfly.channel.MavenArtifact> channelArtifacts = channelSession.resolveMavenArtifacts(mapper.toChannelArtifacts());
 
-    public void resolveAll(List<MavenArtifact> artifacts) throws MavenUniverseException {
-        // use sync until wildfly-channel can handle async search
-        resolveSynchronously(artifacts);
-    }
+            mapper.applyResolution(channelArtifacts);
+        } else {
 
-    private void resolveSynchronously(List<MavenArtifact> artifacts) throws MavenUniverseException {
-        for (MavenArtifact artifact : artifacts) {
-            resolve(artifact);
-        }
-    }
+            List<ArtifactCoordinate> coordinates = toResolvableCoordinates(mapper.toChannelArtifacts());
 
-    private void resolveConcurrently(List<MavenArtifact> artifacts) throws MavenUniverseException {
-        final ExecutorService executorService = Executors.newWorkStealingPool(30);
-        try {
-            List<CompletableFuture<Void>> allPackages = new ArrayList<>();
+            final List<org.wildfly.channel.MavenArtifact> channelArtifacts = channelSession.resolveDirectMavenArtifacts(coordinates);
+
+            mapper.applyResolution(channelArtifacts);
 
             for (MavenArtifact artifact : artifacts) {
-                final CompletableFuture<Void> cf = new CompletableFuture<>();
-                allPackages.add(cf);
-                executorService.submit(() -> {
-                    try {
-                        resolve(artifact);
-                        cf.complete(null);
-                    } catch (Exception e) {
-                        cf.completeExceptionally(e);
+                // workaround for when provisioning doesn't add universe artifacts
+                if (artifact.getPath() == null) {
+                    if (isUniverseOrProducerArtifact(artifact.getArtifactId())) {
+                        org.wildfly.channel.MavenArtifact result = channelSession.resolveMavenArtifact(artifact.getGroupId(), artifact.getArtifactId(), artifact.getExtension(),
```

Comment:
Wouldn't this fail with "no stream found", if universe or producer is not part of channel?

Reply:
Yes, it would. In this case we need to handle a case where those artifacts are in the channel, but not the manifest.

The problem scenario is if the server is installed using maven GA it doesn't initially resolve universe or producer, so they're not listed in the manifest. 
If this manifest is then used to re-provision the server (in 'rollback' or 'restore'),  a FPL will be taken from 'provisioning.xml', which might use the universe notation, which would require universe and producer.

Reply:
OK, thanks for explanation.